py/server/deephaven/time.py

```diff
@@ -155,1885 +135,485 @@ def time_zone_alias_rm(alias: str) -> bool:
-
-    Raises:
-        DHError
-    """
     try:
-        return _JDateTimeUtils.millisToMicros(millis)
+        if dt is None:
+            return None
+        elif isinstance(dt, str):
+            return _JDateTimeUtils.parseLocalDate(dt)
+        elif isinstance(dt, datetime.date) or isinstance(dt, datetime.datetime):
+            return _JLocalDate.of(dt.year, dt.month, dt.day)
+        elif isinstance(dt, np.datetime64):
+            return to_j_local_date(dt.astype(datetime.date))
+        else:
+            raise Exception("Unsupported conversion: " + str(type(dt)) + " -> LocalDate")
     except Exception as e:
         raise DHError(e) from e
 
 
-def seconds_to_micros(seconds: int) -> int:
-    """ Converts seconds to microseconds.
-
-    Args:
-        seconds (int): Seconds to convert.
-
-    Returns:
-        NULL_LONG if the input is NULL_LONG; otherwise the input seconds converted to microseconds.
-
-    Raises:
-        DHError
+def to_j_local_time(dt: Union[None, str, datetime.time, datetime.datetime, np.datetime64]) -> Optional[LocalTime]:
     """
-    try:
-        return _JDateTimeUtils.secondsToMicros(seconds)
-    except Exception as e:
-        raise DHError(e) from e
-
+    Converts a Python string, time, datetime, or numpy datetime64 to a Deephaven local time,
+    which is the time that would be read from a clock and does not have a date or timezone.
 
-def nanos_to_millis(nanos: int) -> int:
-    """ Converts nanoseconds to milliseconds.
+    Time strings can be formatted as 'hh:mm:ss[.nnnnnnnnn]'.
 
     Args:
-        nanos (int): Nanoseconds to convert.
+        dt (Union[None, str, datetime.time, datetime.datetime, np.datetime64]): A Python time, date time, or
+            time string.  If None is provided, None is returned.
 
     Returns:
-        NULL_LONG if the input is NULL_LONG; otherwise the input nanoseconds converted to milliseconds, rounded down.
+        LocalTime
 
     Raises:
         DHError
     """
-    try:
-        return _JDateTimeUtils.nanosToMillis(nanos)
-    except Exception as e:
-        raise DHError(e) from e
-
 
-def micros_to_millis(micros: int) -> int:
-    """ Converts microseconds to milliseconds.
-
-    Args:
-        micros (int): Microseconds to convert.
-
-    Returns:
-        NULL_LONG if the input is NULL_LONG; otherwise the input microseconds converted to milliseconds, rounded down.
-
-    Raises:
-        DHError
-    """
     try:
-        return _JDateTimeUtils.microsToMillis(micros)
+        if dt is None:
+            return None
+        elif isinstance(dt, str):
+            return _JDateTimeUtils.parseLocalTime(dt)
+        elif isinstance(dt, datetime.time) or isinstance(dt, datetime.datetime):
+            return _JLocalTime.of(dt.hour, dt.minute, dt.second, dt.microsecond * 1000)
+        elif isinstance(dt, np.datetime64):
+            # Conversion only supports micros resolution
+            return to_j_local_time(dt.astype(datetime.time))
```

Comment:
Not sure about the lost precision here. How about converting to pandas.Timestamp first? Also should we consider adding pandas.Timestamp to the supported types in those to_j_xxx functions?

Reply:
I do think Pandas timestamps should be supported